### PR TITLE
Adopt structured concurrency

### PR DESCRIPTION
**Pull request by stefanceriu**
This change will refactor most of the code away block based callbacks and GCD and towards swift's new structured concurrency.

### CODE REVIEW

File: ElementX/Sources/Services/Timeline/RoomTimelineController.swift [hunk 1]
```diff
@@ -31,7 +31,10 @@ class RoomTimelineController: RoomTimelineControllerProtocol {
         self.mediaProvider = mediaProvider
         self.memberDetailProvider = memberDetailProvider
         
-        self.timelineProvider.callbacks.sink { [weak self] callback in
+        self.timelineProvider
+            .callbacks
+            .receive(on: DispatchQueue.main)
+            .sink { [weak self] callback in
             guard let self = self else { return }
```
Comment (pixlwave) [inline ElementX/Sources/Services/Timeline/RoomTimelineController.swift:38]: The alignment on this one is slightly off too.

File: ElementX/Sources/Services/Timeline/RoomTimelineController.swift [hunk 2]
```diff
@@ -79,8 +80,13 @@ class RoomTimelineController: RoomTimelineControllerProtocol {
...
     }
     
-    func sendMessage(_ message: String) {
-        timelineProvider.sendMessage(message)
+    func sendMessage(_ message: String) async {
+        switch await timelineProvider.sendMessage(message) {
+        case .success:
+            break
+        case .failure:
+            break
```
Comment (pixlwave) [inline ElementX/Sources/Services/Timeline/RoomTimelineController.swift:86]: Assuming this is for later 🙂

File: ElementX/Sources/Services/Timeline/RoomTimelineController.swift [hunk 3]
```diff
@@ -133,65 +139,69 @@ class RoomTimelineController: RoomTimelineControllerProtocol {
...
-                    return
-                }
-                
-                item.image = image
-                self.timelineItems[index] = item
-                self.callbacks.send(.updatedTimelineItem(timelineItem.id))
-            }
+            item.image = image
+            self.timelineItems[index] = item
+            self.callbacks.send(.updatedTimelineItem(timelineItem.id))
```
Comment (pixlwave) [inline ElementX/Sources/Services/Timeline/RoomTimelineController.swift]: I think this is good now (maybe caught by Sonarcloud anyway)
```suggestion
            timelineItems[index] = item
            callbacks.send(.updatedTimelineItem(timelineItem.id))
```

Comment (pixlwave) [reply ElementX/Sources/Services/Timeline/RoomTimelineController.swift]: Same for the rest below too :)

File: Tools/Scripts/Templates/SimpleScreenExample/ElementX/TemplateSimpleScreenViewModel.swift [hunk 4]
```diff
@@ -16,11 +16,9 @@
 
 import SwiftUI
 
-@available(iOS 14, *)
 typealias TemplateSimpleScreenViewModelType = StateStoreViewModel<TemplateSimpleScreenViewState,
-                                                                  Never,
                                                                   TemplateSimpleScreenViewAction>
```
Comment (pixlwave) [inline Tools/Scripts/Templates/SimpleScreenExample/ElementX/TemplateSimpleScreenViewModel.swift:20]: Could we pop these on the same line now there's only 2? The bad indentation is always annoying on EI.

**Review by pixlwave — COMMENTED** [review]
Looks great 🤩🤩🤩

Do we need changelogs for this repo yet?

File: ElementX/Sources/Services/Media/MediaProvider.swift [hunk 5]
```diff
@@ -77,7 +37,42 @@ struct MediaProvider: MediaProviderProtocol {
...
+    func loadImageFromURL(_ url: String) async -> Result<UIImage, MediaProviderError> {
+        await loadImageFromSource(MediaSource(source: mediaSourceFromUrl(url: url)))
+    }
+        
+    func loadImageFromSource(_ source: MediaSource) async -> Result<UIImage, MediaProviderError> {
+        if let image = imageFromSource(source) {
+            return .success(image)
+        }
+        
+        let cachedImageLoadResult = await withCheckedContinuation({ continuation in
```
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Services/Media/MediaProvider.swift]: I think I already fixed this one 👍

File: ElementX/Sources/Services/Room/Members/MemberDetailsProvider.swift [hunk 6]
```diff
@@ -18,58 +18,38 @@ class MemberDetailProvider: MemberDetailProviderProtocol {
...
     func displayNameForUserId(_ userId: String) -> String? {
-        self.memberDisplayNames[userId]
+        memberDisplayNames[userId]
     }
     
-    func displayNameForUserId(_ userId: String, completion: @escaping (Result<String?, MemberDetailProviderError>) -> Void) {
-        guard let roomProxy = roomProxy else {
-            return
+    func loadDisplayNameForUserId(_ userId: String) async -> Result<String?, MemberDetailProviderError> {
+        if let displayName = displayNameForUserId(userId) {
```
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Services/Room/Members/MemberDetailsProvider.swift:43]: Definitely cached for the life of the application, mainly because hitting the FFI layer for every display name is too heavy. We'll need to find better ways to do that in the future.

File: ElementX/Sources/Services/Room/RoomSummary/RoomSummary.swift [hunk 7]
```diff
@@ -71,81 +71,81 @@ class RoomSummary: RoomSummaryProtocol {
...
+    func loadDetails() async {
         if hasLoadedData {
             return
         }
         
-        loadDisplayName()
-        loadLastMessage()
-        loadAvatar()
+        await withTaskGroup(of: Void.self) { group in
+            group.addTask(priority: .medium) {
```
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Services/Room/RoomSummary/RoomSummary.swift:101]: Because I couldn't figure out a way to have taskGroups with detached tasks I figured it's a good idea to leave them in here to at least give the system an idea of what to do. Now.. all of them will eventually hit a detached task in their respective service so I'm not entirely sure if it's going to make any difference. What do you think?

File: ElementX/Sources/Services/Timeline/RoomTimelineController.swift (hunk 2 above)
Comment (stefanceriu) [reply ElementX/Sources/Services/Timeline/RoomTimelineController.swift:86]: Hah, yes, there's basically no proper error handling anywhere.

File: ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift [hunk 8]
```diff
@@ -64,25 +64,30 @@ class RoomScreenViewModel: RoomScreenViewModelType, RoomScreenViewModelProtocol
...
-        case .itemAppeared(let id):
-            timelineController.processItemAppearance(id)
-        case .itemDisappeared(let id):
-            timelineController.processItemDisappearance(id)
-        case .linkClicked(let url):
-            MXLog.warning("Link clicked: \(url)")
-        case .sendMessage:
-            guard state.bindings.composerText.count > 0 else {
-                return
+        Task {
```
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift:67]: Yep, I think you're right, I'll try it out 👍

File: ElementX/Sources/Services/Authentication/UserSession.swift [hunk 9]
```diff
@@ -75,79 +74,65 @@ class UserSession: ClientDelegate {
...
-                    }
-                    currentRooms.append(RoomProxy(room: sdkRoom, messageFactory: RoomMessageFactory()))
-                case .remove(_, let id, _):
-                    currentRooms.removeAll { $0.id == id }
+        Benchmark.startTrackingForIdentifier("ClientRooms", message: "Fetching available rooms")
+        let sdkRooms = self.client.rooms()
+        Benchmark.endTrackingForIdentifier("ClientRooms", message: "Retrieved \(sdkRooms.count) rooms")
+        
+        Benchmark.startTrackingForIdentifier("ProcessingRooms", message: "Started processing \(sdkRooms.count) rooms")
+        let diff = sdkRooms.map({ $0.id()}).difference(from: currentRooms.map({ $0.id }))
```
[the remark this answers is not in the document]
Comment (pixlwave) [reply ElementX/Sources/Services/Authentication/UserSession.swift]: Fair enough. Could you add an extra space after `id()` though so the closure is balanced?
`let diff = sdkRooms.map({ $0.id() }).difference…`

File: ElementX/Sources/Services/Room/Members/MemberDetailsProvider.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (pixlwave) [reply ElementX/Sources/Services/Room/Members/MemberDetailsProvider.swift:43]: Makes sense.

File: ElementX/Sources/Services/Room/RoomSummary/MockRoomSummary.swift [hunk 10]
```diff
@@ -30,7 +30,7 @@ struct MockRoomSummary: RoomSummaryProtocol {
     
     var avatar: UIImage?
     
-    func loadData() {
+    func loadDetails() {
```
[the remark this answers is not in the document]
Comment (pixlwave) [reply ElementX/Sources/Services/Room/RoomSummary/MockRoomSummary.swift]: I wonder if it only complains the other way, when you define something as `async` which isn't expected in the protocol.

File: ElementX/Sources/Services/Room/RoomSummary/EventBriefFactoryProtocol.swift [hunk 11]
```diff
@@ -8,6 +8,7 @@
 
 import Foundation
 
+@MainActor
 protocol EventBriefFactoryProtocol {
-    func eventBriefForMessage(_ message: RoomMessageProtocol?, completion: @escaping ((EventBrief?) -> Void))
+    func eventBriefForMessage(_ message: RoomMessageProtocol?) async -> EventBrief?
```
[the remark this answers is not in the document]
Comment (pixlwave) [reply ElementX/Sources/Services/Room/RoomSummary/EventBriefFactoryProtocol.swift:13]: For me its at the call site where
```
lastMessage = await eventBriefFactory.eventBrief(for: roomProxy.messages.last)
```
reads much better in my head than
```
lastMessage = await eventBriefFactory.eventBriefForMessage(roomProxy.messages.last)
```
But this is totally unrelated to this PR 🙃

File: ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift [hunk 12]
```diff
@@ -40,7 +38,9 @@ class RoomScreenViewModel: RoomScreenViewModelType, RoomScreenViewModelProtocol
         
         super.init(initialViewState: RoomScreenViewState(roomTitle: roomName ?? "Unknown room 💥", bindings: RoomScreenViewStateBindings(composerText: "")))
         
-        timelineController.callbacks.sink { [weak self] callback in
+        timelineController.callbacks
+            .receive(on: DispatchQueue.main)
+            .sink { [weak self] callback in
             guard let self = self else { return }
```
[the remark this answers is not in the document]
Comment (pixlwave) [reply ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift]: Ok I'm split now. On the wone hand, given there is no actual work I think its probably healthy to let them inherit the default priority and get the detached tasks scheduled more quickly. On the other hand its unlikely this will be called after a user action so `.medium` is probably totally sufficient.

File: ElementX/Sources/Services/Room/RoomSummary/EventBriefFactoryProtocol.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Services/Room/RoomSummary/EventBriefFactoryProtocol.swift:13]: Naming things is hard 😅 

For me `eventBrief(for: )` doesn't fully describe what it is expecting unless I go check the parameter type and it also relies on the on the naming to be on point.

Another option would of course be which would allow you to add more parameters easily.
```swift
brief = await eventBriefFor(message: roomProxy.messages.last)

brief = await eventBriefFor(message: roomProxy.messages.last, identifier: someIdentifier)
```

File: ElementX/Sources/Services/Room/RoomSummary/MockRoomSummary.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Services/Room/RoomSummary/MockRoomSummary.swift]: It does indeed. I assume it's because you can call synchronous methods from async contexts without any problems normally, so not defining it as async doesn't really change anything until you want to run async code from within it as well.

File: ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift (hunk 12 above)
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift]: I think this comment belongs to another conversation.

File: ElementX/Sources/Services/Room/RoomSummary/RoomSummary.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Services/Room/RoomSummary/RoomSummary.swift]: > Ok I'm split now. On the one hand, given there is no actual work I think its probably healthy to let them inherit the default priority and get the detached tasks scheduled more quickly. On the other hand its unlikely this will be called after a user action so `.medium` is probably totally sufficient.

This article explains it well: https://www.hackingwithswift.com/quick-start/concurrency/how-to-control-the-priority-of-a-task

Basically I'll go with this line `not specifying an exact priority is often a good idea because Swift will do The Right Thing` and just remove them for now.

**Review by pixlwave — APPROVED** [review]
LGTM 👏
